File: tools/P爬網站的文章總表_痞客邦.mjs
import axios from "axios";
import * as cheerio from "cheerio";
import fs from "fs";
import path from "path";
import readline from "readline";
import * as XLSX from "xlsx";

const OUT_DIR = "F:\\F\\AI";
const PAGE_SLEEP_MS = 800;        // 翻頁間隔
const ARTICLE_SLEEP_MS = 250;     // 補抓日期間隔（避免太快）
const STOP_EMPTY_PAGES = 3;       // 連續幾頁沒有新文章就停止
const MAX_PAGES_GUARD = 1000;     // 保險上限，避免無限迴圈

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function askUrl() {
  return new Promise((resolve) => {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    rl.question("請輸入要爬取的 PIXNET 網址（首頁或分類頁都可）：\n", (answer) => {
      rl.close();
      resolve((answer || "").trim());
    });
  });
}

function safeFilename(name, maxLen = 120) {
  name = String(name).replace(/[<>:"/\\|?*]/g, "_").trim();
  name = name.replace(/\s+/g, " ");
  if (name.length > maxLen) name = name.slice(0, maxLen).trim();
  return name || "output";
}

function filenameFromUrl(startUrl) {
  const u = new URL(startUrl);
  const host = u.hostname.replace(/\./g, "_");
  const parts = u.pathname.split("/").filter(Boolean);
  const tail = parts.length ? parts.join("_") : "root";
  return safeFilename(`${host}_${tail}`) + ".xlsx";
}

function setPage(url, page) {
  const u = new URL(url);
  u.searchParams.set("page", String(page));
  return u.toString();
}

// 只做「列表頁」抓文章：標題、連結、（若剛好有）日期
function extractPostsFromList($) {
  const rows = [];

  // 以 main/container 優先，避免抓到側欄熱門文章/分類列表
  const root = $("main").length ? $("main") : $("#container").length ? $("#container") : $.root();

  let articles = root.find("article");
  if (articles.length === 0) articles = root.find("h2");

  articles.each((_, el) => {
    const $el = $(el);

    const a = $el.is("h2") ? $el.find("a").first() : $el.find("h2 a").first();
    const link = (a.attr("href") || "").trim();
    const title = a.text().trim();
    if (!link || !title) return;

    // 列表頁日期（若版型沒有就會空，後面會補抓）
    let dt = "";
    const timeEl = $el.find("time[datetime]").first();
    if (timeEl.length) dt = (timeEl.attr("datetime") || "").slice(0, 10);

    rows.push({ 日期: dt, 名稱: title, 網址: link });
  });

  // 以網址去重
  const uniq = new Map();
  for (const r of rows) uniq.set(r.網址, r);
  return [...uniq.values()];
}

// ---------- 補抓日期（進文章頁） ----------
function parseDateToYYYYMMDD(s) {
  if (!s) return "";
  const str = String(s).trim();

  let m = str.match(/(20\d{2})-(\d{1,2})-(\d{1,2})/);
  if (m) {
    const y = m[1];
    const mo = String(parseInt(m[2], 10)).padStart(2, "0");
    const d = String(parseInt(m[3], 10)).padStart(2, "0");
    return `${y}-${mo}-${d}`;
  }

  m = str.match(/(20\d{2})\/(\d{1,2})\/(\d{1,2})/);
  if (m) {
    const y = m[1];
    const mo = String(parseInt(m[2], 10)).padStart(2, "0");
    const d = String(parseInt(m[3], 10)).padStart(2, "0");
    return `${y}-${mo}-${d}`;
  }

  return "";
}

function extractDateFromArticleHtml(html) {
  const $ = cheerio.load(html);

  // 常見 meta
  const metaChecks = [
    ["property", "article:published_time"],
    ["property", "og:published_time"],
    ["name", "pubdate"],
    ["name", "publishdate"],
    ["name", "publish_date"],
    ["name", "date"],
    ["itemprop", "datePublished"],
  ];

  for (const [attr, val] of metaChecks) {
    const tag = $(`meta[${attr}="${val}"]`).first();
    const d = parseDateToYYYYMMDD(tag.attr("content"));
    if (d) return d;
  }

  // time
  const t = $("time").first();
  if (t.length) {
    const d = parseDateToYYYYMMDD(t.attr("datetime") || t.text());
    if (d) return d;
  }

  // regex fallback
  const m = html.match(/\b(20\d{2})[-/](\d{1,2})[-/](\d{1,2})\b/);
  if (m) {
    const y = m[1];
    const mo = String(parseInt(m[2], 10)).padStart(2, "0");
    const d = String(parseInt(m[3], 10)).padStart(2, "0");
    return `${y}-${mo}-${d}`;
  }

  return "";
}

async function fetchArticleDate(axiosInst, url) {
  try {
    const html = (await axiosInst.get(url)).data;
    return extractDateFromArticleHtml(html);
  } catch {
    return "";
  }
}

// ---------- 輸出 xlsx ----------
function writeXlsx(rows, outPath) {
  const data = rows.map((r) => ({
    日期: r.日期 || "",
    名稱: r.名稱 || "",
    網址: r.網址 || "",
  }));

  const ws = XLSX.utils.json_to_sheet(data, { header: ["日期", "名稱", "網址"] });
  ws["!cols"] = [{ wch: 12 }, { wch: 40 }, { wch: 70 }];

  const wb = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(wb, ws, "posts");
  XLSX.writeFile(wb, outPath);
}

async function main() {
  let startUrl = process.argv[2];
  if (!startUrl) startUrl = await askUrl();

  if (!startUrl) {
    console.log("❌ 未輸入網址，結束程式");
    process.exit(1);
  }
  if (!/^https?:\/\//i.test(startUrl)) startUrl = "https://" + startUrl;

  fs.mkdirSync(OUT_DIR, { recursive: true });

  const axiosInst = axios.create({
    headers: { "User-Agent": "Mozilla/5.0" },
    timeout: 30000,
  });

  // ✅ 不依賴 maxPage：一直翻到沒新文章為止
  const seen = new Set();
  const all = [];
  let emptyPages = 0;

  for (let page = 1; page <= MAX_PAGES_GUARD; page++) {
    const pageUrl = setPage(startUrl, page);

    let html = "";
    try {
      html = (await axiosInst.get(pageUrl)).data;
    } catch (e) {
      // 取不到頁面就視為結尾
      console.log(`⚠️ page=${page} 讀取失敗，停止翻頁`);
      break;
    }

    const $ = cheerio.load(html);
    const rows = extractPostsFromList($);

    let added = 0;
    for (const r of rows) {
      if (!seen.has(r.網址)) {
        seen.add(r.網址);
        all.push(r);
        added++;
      }
    }

    console.log(`page=${page} 本頁新增 ${added} 累計 ${all.length}`);

    if (added === 0) emptyPages++;
    else emptyPages = 0;

    if (emptyPages >= STOP_EMPTY_PAGES) {
      console.log(`📌 連續 ${STOP_EMPTY_PAGES} 頁無新文章，停止翻頁`);
      break;
    }

    await sleep(PAGE_SLEEP_MS);
  }

  console.log(`✅ 文章列表蒐集完成：${all.length} 篇`);

  // 補抓日期
  let needFix = all.filter((r) => !r.日期).length;
  console.log(`需要補抓日期：${needFix} 篇（會逐篇進文章頁抓 meta/time）`);

  for (let i = 0; i < all.length; i++) {
    if (all[i].日期) continue;

    const d = await fetchArticleDate(axiosInst, all[i].網址);
    if (d) all[i].日期 = d;

    await sleep(ARTICLE_SLEEP_MS);

    if ((i + 1) % 25 === 0) {
      const left = all.filter((r) => !r.日期).length;
      console.log(`補抓日期進度：${i + 1}/${all.length}，尚未補到日期：${left} 篇`);
    }
  }

  const outFile = filenameFromUrl(startUrl);
  const outPath = path.join(OUT_DIR, outFile);

  writeXlsx(all, outPath);

  console.log(`✅ 完成，共 ${all.length} 筆`);
  console.log(`📄 輸出位置：${outPath}`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: tools/P爬網站的文章總表_痞客邦.py
import os
import re
import time
from urllib.parse import urljoin

import requests
from bs4 import BeautifulSoup
import pandas as pd


OUT_DIR = r"F:\F\AI"
SLEEP_SEC = 0.8  # 爬取間隔


def safe_filename(name: str, max_len: int = 120) -> str:
    # Windows 不可用字元: <>:"/\|?*
    name = re.sub(r'[<>:"/\\|?*]', "_", name).strip()
    name = re.sub(r"\s+", " ", name)
    if len(name) > max_len:
        name = name[:max_len].rstrip()
    return name or "output"


def fetch_soup(session: requests.Session, url: str) -> BeautifulSoup:
    r = session.get(url, timeout=30)
    r.raise_for_status()
    return BeautifulSoup(r.text, "html.parser")


def extract_posts(soup: BeautifulSoup):
    rows = []

    articles = soup.select("article")
    if not articles:
        articles = soup.select("h2")

    for a in articles:
        title_a = a.select_one("h2 a") if a.name != "h2" else a.select_one("a")
        if not title_a or not title_a.get("href"):
            continue

        title = title_a.get_text(strip=True)
        link = title_a["href"].strip()

        dt = ""
        time_el = a.select_one("time[datetime]") if a.name != "h2" else None
        if time_el and time_el.get("datetime"):
            dt = time_el["datetime"][:10]
        else:
            time_any = a.select_one("time") if a.name != "h2" else None
            if time_any:
                dt = time_any.get_text(strip=True)

        rows.append({
            "日期": dt,
            "名稱": title,
            "網址": link
        })

    uniq = {}
    for r in rows:
        uniq[r["網址"]] = r
    return list(uniq.values())


def find_next_page(soup: BeautifulSoup, current_url: str):
    a = soup.select_one('a[rel="next"]')
    if a and a.get("href"):
        return urljoin(current_url, a["href"])

    for cand in soup.select("a"):
        if "next" in cand.get_text(strip=True).lower() and cand.get("href"):
            return urljoin(current_url, cand["href"])
    return None


def main():
    start_url = input("請輸入要爬取的 EZQuant Blog 網址：\n").strip()
    if not start_url:
        print("❌ 未輸入網址，結束程式")
        return

    os.makedirs(OUT_DIR, exist_ok=True)

    with requests.Session() as s:
        s.headers.update({
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"
        })

        first_soup = fetch_soup(s, start_url)

        page_title = first_soup.title.get_text(strip=True) if first_soup.title else "blog"
        file_name = safe_filename(page_title) + ".xlsx"
        out_path = os.path.join(OUT_DIR, file_name)

        all_rows = []
        url = start_url
        seen = set()

        while url and url not in seen:
            seen.add(url)
            soup = first_soup if url == start_url else fetch_soup(s, url)

            all_rows.extend(extract_posts(soup))
            url = find_next_page(soup, url)

            time.sleep(SLEEP_SEC)

        uniq = {}
        for r in all_rows:
            uniq[r["網址"]] = r

        df = pd.DataFrame(
            uniq.values(),
            columns=["日期", "名稱", "網址"]
        )

        df.to_excel(out_path, index=False)
        print(f"✅ 完成，共 {len(df)} 筆")
        print(f"📄 輸出位置：{out_path}")


if __name__ == "__main__":
    main()
